### PR TITLE
[MNG-6819] Fix NullPointerException in StringVisitorModelInterpolator$ModelVisitor

### PR DESCRIPTION
**Pull request by sadhen**
## Desc

Here is a clue to reproduce the NPE.
```
            <plugin>
                <artifactId>maven-antrun-plugin</artifactId>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <configuration>
                            <tasks>
                                <unzip
                                    src="${project.build.directory}/${project.build.finalName}.${project.packaging}"
                                    dest="${project.build.directory}/xxx" />
                            </tasks>
                        </configuration>
                        <goals>
                            <goal>run</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

`project.build.finalName` is the previous value, and the interpolated value is null.

Fix NullPointerException.

null is not allowed as a Hashtable value.

```
$ mvn clean compile -e  # on a private project
[ERROR] NullPointerException
java.lang.NullPointerException
    at java.util.Hashtable$Entry.setValue (Hashtable.java:1286)
    at org.apache.maven.model.interpolation.StringVisitorModelInterpolator$ModelVisitor.visit (StringVisitorModelInterpolator.java:1429)
    at org.apache.maven.model.interpolation.StringVisitorModelInterpolator$ModelVisitor.visit (StringVisitorModelInterpolator.java:1027)
    at org.apache.maven.model.interpolation.StringVisitorModelInterpolator$ModelVisitor.visit (StringVisitorModelInterpolator.java:170)
    at org.apache.maven.model.interpolation.StringVisitorModelInterpolator.interpolateModel (StringVisitorModelInterpolator.java:107)
    at org.apache.maven.model.building.DefaultModelBuilder.interpolateModel (DefaultModelBuilder.java:789)
    at org.apache.maven.model.building.DefaultModelBuilder.build (DefaultModelBuilder.java:393)
    at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom (DefaultArtifactDescriptorReader.java:292)
    at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor (DefaultArtifactDescriptorReader.java:171)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor (DefaultDependencyCollector.java:541)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult (DefaultDependencyCollector.java:524)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:412)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.doRecurse (DefaultDependencyCollector.java:509)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:461)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:365)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:352)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies (DefaultDependencyCollector.java:254)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies (DefaultRepositorySystem.java:284)
    at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve (DefaultProjectDependenciesResolver.java:169)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies (LifecycleDependencyResolver.java:243)
    at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies (LifecycleDependencyResolver.java:147)
    at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved (MojoExecutor.java:248)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:202)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

## Checklist
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by michael-o** [comment]
This would require an JIRA issue because you have found an obvious bug. How did you find it?

File: maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java [hunk 1]
```diff
@@ -1424,7 +1420,7 @@ private void visit( Properties properties )
                     {
                         String value = (String) v;
                         String inter = interpolate( value );
-                        if ( value != inter )
+                        if ( value != inter && Objects.nonNull( inter ) )
```
Comment (Tibor17) [inline maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java]: `nonNull` is used in Lambda with Method Reference and not like this.
That\s the reason why `Objects.nonNull` was introduced in Java 1.8.
Simple is better. Therefore `!= null`.

**Comment by Tibor17** [comment]
@sadhen 

Is this NPE a blocker?

@michael-o 
We cannot release 3.6.4 due to https://github.com/apache/maven/commit/a4317947e60ec2573827b017b82e1556610d8074
Only a branch is possible.

**Comment by michael-o** [comment]
> 
> 
> @sadhen
> 
> Is this NPE a blocker?
> 
> @michael-o
> We cannot release 3.6.4 due to [a431794](https://github.com/apache/maven/commit/a4317947e60ec2573827b017b82e1556610d8074)
> Only a branch is possible.

we can branch off 3.6.x.

**Comment by sadhen** [comment]
@michael-o 

I have attached the stack trace.

Sorry for the rush pull request, since it is Friday night.

I tried to write a reproducer but failed for several attempts.

**Comment by rfscholte** [comment]
I have my doubt that this is the bug. I would love to see a failing test first.

**Comment by slachiewicz** [comment]
yes, it's bug - simple pom.xml to reproduce the problem locally

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
 <modelVersion>4.0.0</modelVersion>

<groupId>test</groupId>
 <artifactId>test</artifactId>
 <version>1.0</version>

<properties>
 <aa>${bb}</aa>
 <bb>${aa}</bb>
</properties>

</project>

```

**Comment by rfscholte** [comment]
Can you make a unittest for it?

**Comment by slachiewicz** [comment]
sure - https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=9c46b06bfc080ad51d2d5626114f47293693d05c

File: maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java [hunk 2]
```diff
@@ -1424,7 +1424,7 @@ private void visit( Properties properties )
                     {
                         String value = (String) v;
                         String inter = interpolate( value );
-                        if ( value != inter )
+                        if ( value != inter && inter != null )
```
Comment (hepin1989) [inline maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java:1427]: Better to add logging for the interpolated value is null.

Comment (michael-o) [reply maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java:1427]: But at most at debug level or even trace.

Comment (hepin1989) [reply maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java:1427]: I encounter this issue too, have to make use of maven 3.5 , failed on 3.6

**Comment by DarioRomano** [comment]
I've also run into this issue when I added an xtext dependency to my build. It seems to be having troubles resolving transitive dependencies.

File: maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java (hunk 2 above)
Comment (slachiewicz) [reply maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringVisitorModelInterpolator.java:1427]: there is no need to log any - null is the normal value returned from `interpolate()` function

**Comment by olamy** [comment]
easy null check. Thanks for that!

**Comment by jdillon** [comment]
Any reason why this was not back-ported to 3.8.x?

Also this change only guards `visit( Properties properties )` why does it not guard `visit( List<String> list )` in the same way?  Isn't the spirit of maven interpolation to leave the value asis if it couldn't interpolate anything?  Though my memory on what its supposed to be done is a bit fuzzy.

**Comment by michael-o** [comment]
> 
> 
> Any reason why this was not back-ported to 3.8.x?
> 
> Also this change only guards `visit( Properties properties )` why does it not guard `visit( List<String> list )` in the same way? Isn't the spirit of maven interpolation to leave the value asis if it couldn't interpolate anything? Though my memory on what its supposed to be done is a bit fuzzy.

3.8.x will be assessed after 4.0.0-alpha-1.

**Comment by michael-o** [comment]
> 
> 
> Any reason why this was not back-ported to 3.8.x?
> 
> Also this change only guards `visit( Properties properties )` why does it not guard `visit( List<String> list )` in the same way? Isn't the spirit of maven interpolation to leave the value asis if it couldn't interpolate anything? Though my memory on what its supposed to be done is a bit fuzzy.

Please leave a note in the ticket.

**Comment by jira-importer** [comment]
Resolve #8649

**Pull request closed without merging**